c_src/Tensorflex.c

```
@@ -150,29 +119,47 @@ static ERL_NIF_TERM new_op(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
   return op_desc;
 }
 
-static ERL_NIF_TERM graph_import_graph_def(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
+static ERL_NIF_TERM read_graph(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
 {
-  TF_Graph **graph;
-  enif_get_resource(env, argv[0], graph_resource, (void *) &graph);
+  ErlNifBinary filepath;
+  enif_inspect_binary(env,argv[0], &filepath);
+
+  char* file = enif_alloc(filepath.size+1);
+  memset(file, 0, filepath.size+1);
+  memcpy(file, (void *) filepath.data, filepath.size);
+
+  FILE *f = fopen(file, "rb");
+  fseek(f, 0, SEEK_END);
+  long fsize = ftell(f);
+  fseek(f, 0, SEEK_SET);
 
-  TF_Buffer **buffer;
-  enif_get_resource(env, argv[1], buffer_resource, (void *) &buffer);
+  void* data = malloc(fsize);
+  fread(data, fsize, 1, f);
+  fclose(f);
 
-  TF_ImportGraphDefOptions **graph_opts;
-  enif_get_resource(env, argv[2], graph_opts_resource, (void *) &graph_opts);
+  TF_Buffer* buf = TF_NewBuffer();
+  buf->data = data;
+  buf->length = fsize;
+  buf->data_deallocator = free_buffer;
 
   TF_Status* status = TF_NewStatus();
+  TF_ImportGraphDefOptions *graph_opts = TF_NewImportGraphDefOptions();
+  TF_Graph *graph = TF_NewGraph();
   
-  TF_GraphImportGraphDef(*graph, *buffer, *graph_opts, status);
+  TF_GraphImportGraphDef(graph, buf, graph_opts, status);
   if (TF_GetCode(status) != TF_OK) {
-    fprintf(stderr, "ERROR: Unable to import graph %s", TF_Message(status));
-    return 1;
+    return enif_make_tuple2(env,enif_make_atom(env,"error"),enif_make_string(env, "Unable to import graph", ERL_NIF_LATIN1));
```

Just to clarify, an Erlang string is not the same as an Elixir string. So we need to use the binary API here.

Also, note that we can return a better error, so let's do that. Here are all of the possible error values from [Tensorflow](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/c/c_api.h#L133-L151). The idea is to convert each of them into an atom. So if we get `TF_NOT_FOUND`, we should return `{:error, :not_found}`. This will probably be very common, so we can encapsulate it in a function.

---

Noted. I completely agree-- I didn't realize I was returning char lists. Now I've shifted to `enif_make_binary()` wherever required. Also added this and the TF error codes to the latest PR.